python/cudf/cudf/_lib/join.pyx

```diff
@@ -222,3 +201,30 @@ cpdef join(Table lhs,
         index = None
     data_ordered_dict = OrderedDict(zip(result_col_names, all_cols_py))
     return Table(data=data_ordered_dict, index=index)
+
+
+def compute_result_col_names(lhs, rhs, how):
+    """
+    Determine the names of the data columns in the result of
+    a libcudf join, based on the original left and right frames
+    as well as the type of join that was performed.
+    """
+    if how in ("left", "inner", "outer"):
+        # the result cols are all the left columns (incl. common ones)
+        # + all the right columns (excluding the common ones)
+        result_col_names = [None] * len(
+            lhs._data.keys() | rhs._data.keys()
+        )
+        ix = 0
+        for name in lhs._data.keys():
+            result_col_names[ix] = name
+            ix += 1
+        for name in rhs._data.keys():
+            if name not in lhs._data.keys():
+                nom = name
+                result_col_names[ix] = nom
+                ix += 1
+    elif how in ("leftsemi", "leftanti"):
+        # the result columns are just all the left columns
+        result_col_names = list(lhs._data.keys())
+    return result_col_names
```

Comment:
Perhaps consider the pythonic equivalent ? 

```code
    if how in ("left", "inner", "outer", "leftsemi", "leftanti"):
        a = list(lhs.keys())
        b = [] if how not in ("left", "inner", "outer") else \
            [k for k in rhs.keys() if k not in lhs.keys()]
        return a+b
```

Quite a substantial patch by the way, thanks @brandon-b-miller !

Reply:
much better thank you! :)